strawberryfields/tdm/tdmprogram.py

```diff
@@ -83,63 +84,134 @@ def validate_measurements(circuit, N):
     defined to be measured in the circuit, instead of being spread over a larger
     number of modes due to the mode-shifting occurring in :class:`~.TDMProgram`.
 
+    The function iterates through the unrolled circuit and populates, and returns, a new samples
+    dictionary with the shape ``{spatial mode: (shots, timebins)}``. E.g., this unrolled circuit:
+
+    .. code-block:: pycon
+
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 0, spatial mode 0  (sample 0)
+        MeasureHomodyne(0) | (q[2])  # shot 0, timebin 0, spatial mode 2  (sample 1)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 0, timebin 1, spatial mode 0  (sample 2)
+        MeasureHomodyne(0) | (q[1])  # shot 0, timebin 1, spatial mode 2  (sample 3)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 1, timebin 0, spatial mode 0  (sample 4)
+        MeasureHomodyne(0) | (q[2])  # shot 1, timebin 0, spatial mode 2  (sample 5)
+        ...
+        MeasureHomodyne(0) | (q[0])  # shot 1, timebin 1, spatial mode 0  (sample 6)
+        MeasureHomodyne(0) | (q[1])  # shot 1, timebin 1, spatial mode 2  (sample 7)
+
+    would return the dictionary
+
+    .. code-block:: pycon
+
+        {
+            0: [[(sample 0), (sample 2)],
+                [(sample 4), (sample 6)]],
+            2: [[(sample 1), (sample 3)],
+                [(sample 5), (sample 7)]],
+        }
+
     Args:
         all_samples (dict[int, list]): the raw measured samples
         modes (Sequence[int]): the modes that are measured in the circuit
         N (Sequence[int]): the number of concurrent modes per belt/spatial modes
+        timebins (int): the number of timebins/temporal modes in the program per shot
 
     Returns:
-        new_samples (dict[int, list]): the re-shaped samples
+        dict[int, list]: the re-shaped samples, where each key correspond to a spatial
+            mode and the values have shape ``(shots, timebins)``
     """
+    # calculate the total number of samples and the order in which they were measured
     num_of_values = len([i for j in all_samples.values() for i in j])
-    mode_order = _get_mode_order(num_of_values, N)
+    mode_order = _get_mode_order(num_of_values, N, timebins)
 
-    # go backwards through all_samples and add them into the correct mode
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
+    t = 0
     for i, m in enumerate(mode_order):
         idx = modes[i % len(N)]
         if idx not in new_samples:
-            new_samples[idx] = []
-        new_samples[idx].append(all_samples[m].pop(0))
-    return new_samples
+            new_samples[idx] = [[] for _ in range(timebins)]
+        new_samples[idx][t].append(all_samples[m].pop(0)[0])
+
+        # populate each spatial mode in one timebin before moving on to the next timebin
+        # when each timebin has been filled, move to the next shot
+        if (i + 1) % len(N) == 0:
```

Comment:
(Minor suggestion): naming the condition could make the operation more descriptive
```suggestion
        # when each timebin has been filled, move to the next shot
        last_mode_in_timebin = (i + 1) % len(N) == 0
        if last_mode_in_timebin:
```